core/types/subtyping.cc

```diff
@@ -1257,32 +1288,81 @@ bool isSubTypeUnderConstraintSingle(const GlobalState &gs, TypeConstraint &const
                         int i = -1;
                         for (auto &el2 : a2->elems) {
                             ++i;
-                            result = Types::isSubTypeUnderConstraint(gs, constr, a1.elems[i], el2, mode);
-                            if (!result) {
-                                break;
+                            auto subCollector = errorDetailsCollector.newCollector();
+                            if (!Types::isSubTypeUnderConstraint(gs, constr, a1.elems[i], el2, mode, subCollector)) {
+                                result = false;
+                                if constexpr (std::is_same<T, ErrorDetailsCollector>::value) {
+                                    // TODO(neil): once we pass in whether this a covariant or contravariant call,
+                                    // we should reword it in the "expected ... but got ..." form
+                                    auto message = ErrorColors::format("`{}` is not a subtype of `{}` for item `{}`",
+                                                                       a1.elems[i].show(gs), el2.show(gs), i);
```

Comment:
circling back after seeing this error in context, I also think it's important to mention tuple in the error message, as it can be easy to miss that. So my favorite is `for tuple index 0`